### ISSUE TITLE
Multiple TestServer instances seem to be sharing a context / seem to be pointing to the same server.

### ISSUE THREAD

**Issue by Yablargo** [aspnet/Home#1541]
Hello. I have a number of Integration Tests that use a shared TestServer instance and a TestStartup class extended from Startup that contains a Seed() method and uses InMemoryDatabase.

In dnx / RC1, all of my tests passed without issues. 

I have a subset of the tests that launch a separate (clean) server and tests initial database values before the rest start fiddling with things.

When I upgraded to RC2, I use this to launch my test server

`Server = new TestServer(new WebHostBuilder().UseContentRoot(System.IO.Directory.GetCurrentDirectory()).UseStartup<TestStartup>());`

But what I find now is that if i run my tests by themselves, they pass, but when I run them together simultaniously, my separate TestServers get the same contexts.

Is there a way I can scope my Context such that I get a new one for each TestServer instance?
I'm getting my context in the Seed() -- called from Configure() like so:

`var db = app.ApplicationServices.GetService<ApplicationDbContext>();`

and it seems somewhat obvious that my app already has a context and returns the same instance

**Comment by Yablargo** [comment]
I guess a better question would be: Can I spawn a completely separate TestServer to check the initial values? Depending on which tests run first, this fails or succeeds

**Comment by davidfowl** [comment]
Each `WebHostBuilder` instance should be isolated. If you can provide a basic repro application we can take a look. 

/cc @Tratcher @JunTaoLuo

**Comment by Yablargo** [comment]
Allright, Will do. Just put it in a repo here and link it?

**Comment by Tratcher** [comment]
Yes please

**Comment by Yablargo** [comment]
https://github.com/Yablargo/SharedContextRepro

**Comment by Tratcher** [comment]
Sharing a TestServer instance statically across tests is not a good idea for unit tests. There's too much of a risk that the tests can have side-effects on eachother. Note xunit can execute tests in parallel, but by default it will only do this for tests in different classes.

I doubt you're getting the same DbContext instance, but you may be connecting to the same in-memory database.  @bricelam does UseInMemoryDatabase create a single static instance in memory that may be inadvertently shared across unit tests?

**Comment by Yablargo** [comment]
@Tratcher the seed data in reality is setup so that all of the tests don't overlap with eachother. I do quite a lot of crunching across a lot of tests and got it so that none of them step on eachother's toes, with the exception of the few that want to spawn their base instance and check the fresh data. 

That makes more sense on your 2nd point -- any way I can get a 2nd instance?

I will say that, if I had it to do all over I'd have separated them out completely, but i spent a LOT of time getting all of the integration tests to work in parallel without any overlap

**Comment by JunTaoLuo** [comment]
Yup, this seems to be an issue with EF.

> "Typically, EF creates a single IServiceProvider for all contexts of a given type in an AppDomain - meaning all context instances share the same InMemory database instance."

Seems like EF's in memory database is by default only one instance per AppDomain. Just as a sanity check, I converted the app to 1.0.0 (which has this fix https://github.com/aspnet/EntityFramework/issues/3253) allowing me to do something like:
```
                 options.UseInMemoryDatabase($"db{System.Threading.Thread.CurrentThread.ManagedThreadId}", o =>
                 {
                     Console.WriteLine("Use InMemory");
                 }));
```
This allows each test to have a unique database specified by its name. With this fix, all tests passed when run in parallel.

https://docs.efproject.net/en/latest/miscellaneous/testing.html seems to have some documentation for writing tests using EF's InMemoryDatabase so I'd take a look there.

**Comment by GeorgDangl** [comment]
Here's how I do it in RC2:

    var services = new ServiceCollection();

    services.AddEntityFrameworkInMemoryDatabase()
        .AddDbContext<MyDbContext>(x => x.UseInMemoryDatabase().UseInternalServiceProvider(services.BuildServiceProvider()));

    var serviceProvider = services.BuildServiceProvider();
    Context = serviceProvider.GetRequiredService<MyDbContext>();`

Special note here to `.UseInternalServiceProvider(services.BuildServiceProvider())`. It's comment says that "_EF will create and manage a service provider if none is specified_", which I think means that one service provider will be shared for the current AppDomain, meaning all InMemory databases get the same data store.

Here, I'm essentially creating a new service provider for the EF dbContext on each class initialization so I have separate data storages accross tests.

**Comment by Yablargo** [comment]
@GeorgDangl Thanks. I'm trying this change now and will be pretty stoked if it fixes everything. I probably won't be moving to 1.0 for a few weeks while we still tie down everything stability-wise

**Comment by Yablargo** [comment]
I'm using RTM now, and while this has fixed my 'shared-instance' issue, there are definately some bugs floating around still with this.

The upshot: **most** of my tests run synchronously.

The negative: The database instances appear to share the same primary key sequence between the various databases that are created at the same time. 

I have a twofold problem at the moment:

if in my seed I use:

```
db.Things.Add(new Thing{whatever="foo"});

```

the new 'Thing' will have a seemingly random ID such as 5 or 6, depending on the number of concurrent databases exist, so I can't expect a particular version of my seed data.

If I explicitly set an Id value:

```
db.Things.Add(new Thing{whatever="foo", ID = 1});

```

When my controllers normally do a 

```
db.Things.add(....) it fails, and complains that there is an id conflict on ID 1, since my server's identity counter is at 1 (it doesn't check to see if this ID already exists)

```

**Comment by Yablargo** [comment]
After some fixes yesterday, nearly everything works, with the quirk of the shared-sequences referred in the previous post.

What I ended up doing is seeding all of my values at arbitrary ID's (e.g. 1000,1001, etc), and the auto inserts start at 1,2,3 and there is not a conflict.

**Comment by ielcoro** [comment]
I see the same behaviour as Yablargo with sequences. Looks like they are all shared between InMemory instances.

**Comment by aspnet-hello** [comment]
This issue is being closed because it has not been updated in 3 months.

We apologize if this causes any inconvenience. We ask that if you are still encountering this issue, please log a new issue with updated information and we will investigate.
